File: mcdowell/php/ch9-05.php
<?php

function permutation($text, $index) {
    if ($index == 0) return [$text{0}];

    $ans = permutation($text, $index - 1);
    $newTextValue = $text{$index};

    $newAns = $ans;

    foreach($ans as $ansItem) {
        for($i = 0; $i <= strlen($ansItem); $i ++) {
            $newAnsItem = substr($ansItem, 0, $i) . $newTextValue . substr($ansItem, $i);
            $newAns[] = $newAnsItem;
        }
    }

    return $newAns;
}

print_r(permutation('abc', 2));

File: mcdowell/python/ch9-05.py
def permutation(text, index):
    if index == 0:
        return [text[0]]

    ans = permutation(text, index - 1)
    newAns = ans[:]
    newLetter = text[index]

    for item in ans:
        for i in range(0, len(item) + 1):
            newItem = item[0:i] + newLetter + item[i:]
            newAns.append(newItem)


    return newAns



print permutation('abc', 2)